File: APP03022-mrch.backend.somx.fiscal-api/src/main/java/com/sodimac/fiscal/api/service/MessageCatalogService.java
package com.sodimac.fiscal.api.service;

import com.sodimac.fiscal.api.model.enums.FiscalMessageCode;
import com.sodimac.fiscal.api.model.enums.FiscalWarningCode;

/**
 * Servicio para gestión centralizada de mensajes del sistema fiscal.
 *
 * Este servicio actúa como una capa de abstracción para obtener mensajes
 * desde un catálogo centralizado. Funciona con un switch de configuración:
 * - catalogs.api.enabled=true: Consulta catalogos-api para mensajes multi-idioma
 * - catalogs.api.enabled=false: Usa el mensaje fallback del enum
 *
 * Los mensajes de negocio (BUS###, ERR###) se consultan en BD cuando está habilitado.
 * Las validaciones de entrada y excepciones genéricas no pasan por este servicio.
 *
 * @author Sodimac Tech Team
 * @version 2.0
 * @since 2025
 */
public interface MessageCatalogService {

    /**
     * Obtiene el mensaje por código desde el catálogo.
     * Consulta BD si catalogs.api.enabled=true, sino usa fallback del enum.
     *
     * @param messageCode Código de mensaje (ERR###, BUS###, etc.)
     * @return Mensaje del catálogo o fallback
     */
    String getMessage(FiscalMessageCode messageCode);

    /**
     * Obtiene el mensaje formateado con parámetros adicionales.
     *
     * @param messageCode Código de mensaje
     * @param params Parámetros adicionales
     * @return Mensaje formateado
     */
    String getMessage(FiscalMessageCode messageCode, Object... params);

    /**
     * Obtiene el mensaje de advertencia por código.
     *
     * @param warningCode Código de advertencia
     * @return Mensaje de advertencia
     */
    String getWarningMessage(FiscalWarningCode warningCode);

    /**
     * Obtiene el mensaje de advertencia formateado con parámetros adicionales.
     *
     * @param warningCode Código de advertencia
     * @param params Parámetros adicionales
     * @return Mensaje formateado
     */
    String getWarningMessage(FiscalWarningCode warningCode, Object... params);

    /**
     * Obtiene el código del mensaje.
     *
     * @param messageCode Enumeración de mensaje
     * @return Código del mensaje (ej: BUS034, ERR001)
     */
    String getCode(FiscalMessageCode messageCode);

    /**
     * Obtiene el código de advertencia.
     *
     * @param warningCode Enumeración de advertencia
     * @return Código de advertencia (ej: WRN009)
     */
    String getWarningCode(FiscalWarningCode warningCode);

    /**
     * Construye un mensaje completo con código y descripción.
     *
     * @param messageCode Código de mensaje
     * @return Mensaje completo [CÓDIGO] Descripción
     */
    String getFullMessage(FiscalMessageCode messageCode);

    /**
     * Construye un mensaje completo con código y descripción.
     *
     * @param warningCode Código de advertencia
     * @return Mensaje completo [CÓDIGO] Descripción
     */
    String getFullWarningMessage(FiscalWarningCode warningCode);

    /**
     * Lanza una FiscalException con el código de mensaje especificado.
     * Consulta el catálogo antes de lanzar si está habilitado.
     *
     * @param messageCode Código de mensaje
     * @throws com.sodimac.fiscal.api.exception.FiscalException
     */
    void throwException(FiscalMessageCode messageCode);

    /**
     * Lanza una FiscalException con el código de mensaje y parámetros para formatear.
     * Los parámetros reemplazan placeholders {0}, {1}, etc. en el mensaje.
     *
     * Ejemplo de uso:
     * <pre>
     * // Mensaje: "El RFC del receptor {0} no acepta documentos fiscales"
     * messageCatalog.throwExceptionWithParams(FiscalMessageCode.BUS002, rfcReceptor);
     * </pre>
     *
     * @param messageCode Código de mensaje
     * @param params Parámetros para formatear el mensaje ({0}, {1}, etc.)
     * @throws com.sodimac.fiscal.api.exception.FiscalException
     */
    void throwExceptionWithParams(FiscalMessageCode messageCode, Object... params);

    /**
     * Lanza una FiscalException con el código de mensaje e información adicional.
     * Consulta el catálogo antes de lanzar si está habilitado.
     *
     * @param messageCode Código de mensaje
     * @param additionalInfo Información adicional del contexto
     * @throws com.sodimac.fiscal.api.exception.FiscalException
     */
    void throwException(FiscalMessageCode messageCode, String additionalInfo);

    /**
     * Lanza una FiscalException con el código de mensaje y causa original.
     * Consulta el catálogo antes de lanzar si está habilitado.
     *
     * @param messageCode Código de mensaje
     * @param cause Excepción original
     * @throws com.sodimac.fiscal.api.exception.FiscalException
     */
    void throwException(FiscalMessageCode messageCode, Throwable cause);

    /**
     * Lanza una FiscalException con el código de mensaje, información adicional y causa.
     * Consulta el catálogo antes de lanzar si está habilitado.
     *
     * @param messageCode Código de mensaje
     * @param additionalInfo Información adicional del contexto
     * @param cause Excepción original
     * @throws com.sodimac.fiscal.api.exception.FiscalException
     */
    void throwException(FiscalMessageCode messageCode, String additionalInfo, Throwable cause);

    // ========== MÉTODOS CREATE EXCEPTION (para uso con throw) ==========

    /**
     * Crea una FiscalException sin lanzarla.
     * Útil para usar con 'throw' y evitar 'return null' después de throwException().
     *
     * Uso recomendado:
     * <pre>
     * throw messageCatalog.createException(FiscalMessageCode.BUS034);
     * </pre>
     *
     * @param messageCode Código de mensaje
     * @return FiscalException lista para lanzar
     */
    com.sodimac.fiscal.api.exception.FiscalException createException(FiscalMessageCode messageCode);

    /**
     * Crea una FiscalException con parámetros para formatear sin lanzarla.
     * Los parámetros reemplazan placeholders {0}, {1}, etc. en el mensaje.
     *
     * Uso recomendado:
     * <pre>
     * // Mensaje: "El RFC del receptor {0} no acepta documentos fiscales"
     * throw messageCatalog.createExceptionWithParams(FiscalMessageCode.BUS002, rfcReceptor);
     * </pre>
     *
     * @param messageCode Código de mensaje
     * @param params Parámetros para formatear el mensaje ({0}, {1}, etc.)
     * @return FiscalException lista para lanzar con mensaje formateado
     */
    com.sodimac.fiscal.api.exception.FiscalException createExceptionWithParams(FiscalMessageCode messageCode, Object... params);

    /**
     * Crea una FiscalException con información adicional sin lanzarla.
     *
     * @param messageCode Código de mensaje
     * @param additionalInfo Información adicional del contexto
     * @return FiscalException lista para lanzar
     */
    com.sodimac.fiscal.api.exception.FiscalException createException(FiscalMessageCode messageCode, String additionalInfo);

    /**
     * Crea una FiscalException con causa original sin lanzarla.
     *
     * @param messageCode Código de mensaje
     * @param cause Excepción original
     * @return FiscalException lista para lanzar
     */
    com.sodimac.fiscal.api.exception.FiscalException createException(FiscalMessageCode messageCode, Throwable cause);

    /**
     * Crea una FiscalException con información adicional y causa sin lanzarla.
     *
     * @param messageCode Código de mensaje
     * @param additionalInfo Información adicional del contexto
     * @param cause Excepción original
     * @return FiscalException lista para lanzar
     */
    com.sodimac.fiscal.api.exception.FiscalException createException(FiscalMessageCode messageCode, String additionalInfo, Throwable cause);

    // ========== MÉTODOS DEPRECADOS PARA COMPATIBILIDAD ==========

    /**
     * @deprecated Usar getMessage(FiscalMessageCode) en su lugar
     */
    @Deprecated
    default String getErrorMessage(FiscalMessageCode messageCode) {
        return getMessage(messageCode);
    }

    /**
     * @deprecated Usar getMessage(FiscalMessageCode, Object...) en su lugar
     */
    @Deprecated
    default String getErrorMessage(FiscalMessageCode messageCode, Object... params) {
        return getMessage(messageCode, params);
    }

    /**
     * @deprecated Usar getCode(FiscalMessageCode) en su lugar
     */
    @Deprecated
    default String getErrorCode(FiscalMessageCode messageCode) {
        return getCode(messageCode);
    }

    /**
     * @deprecated Usar getFullMessage(FiscalMessageCode) en su lugar
     */
    @Deprecated
    default String getFullErrorMessage(FiscalMessageCode messageCode) {
        return getFullMessage(messageCode);
    }

    /**
     * @deprecated Usar throwException(FiscalMessageCode) en su lugar
     */
    @Deprecated
    default void throwError(FiscalMessageCode messageCode) {
        throwException(messageCode);
    }

    /**
     * @deprecated Usar throwException(FiscalMessageCode, String) en su lugar
     */
    @Deprecated
    default void throwError(FiscalMessageCode messageCode, String additionalInfo) {
        throwException(messageCode, additionalInfo);
    }

    /**
     * @deprecated Usar throwException(FiscalMessageCode, Throwable) en su lugar
     */
    @Deprecated
    default void throwError(FiscalMessageCode messageCode, Throwable cause) {
        throwException(messageCode, cause);
    }

    /**
     * @deprecated Usar throwException(FiscalMessageCode, String, Throwable) en su lugar
     */
    @Deprecated
    default void throwError(FiscalMessageCode messageCode, String additionalInfo, Throwable cause) {
        throwException(messageCode, additionalInfo, cause);
    }
}

File: APP03022-mrch.backend.somx.util-api/src/services/messageCatalog.service.ts
// src/services/messageCatalog.service.ts
/**
 * Servicio de catalogo de mensajes para utils-api
 * Similar a MessageCatalogService en fiscal-api
 *
 * ARQUITECTURA:
 * - Si CATALOGOS_API_ENABLED=true: Consulta catalogos-api para mensajes multi-idioma
 * - Si CATALOGOS_API_ENABLED=false: Usa mensajes fallback del enum
 *
 * MULTI-IDIOMA:
 * El idioma se pasa como parametro o se usa el default de configuracion.
 */

import { catalogosClient } from '../clients/catalogos.client.js';
import catalogosConfig from '../config/catalogos.config.js';
import { UtilsMessageCode, UtilsMessageFallback } from '../enums/UtilsMessageCode.js';
import { UtilsException } from '../exceptions/UtilsException.js';

/**
 * Idiomas soportados
 */
export enum LanguageId {
    ES = 1,
    EN = 2,
    PT = 3,
}

/**
 * Servicio de mensajes centralizado
 */
class MessageCatalogService {
    /**
     * Obtiene el mensaje por codigo desde el catalogo
     * @param code Codigo de mensaje
     * @param langId ID del idioma (opcional)
     * @returns Mensaje del catalogo o fallback
     */
    async getMessage(code: UtilsMessageCode, langId?: LanguageId): Promise<string> {
        const lang = langId ?? catalogosConfig.defaultLangId;

        // Intentar obtener de catalogos-api
        if (catalogosConfig.enabled) {
            const message = await catalogosClient.getMessage(code, lang);
            if (message) {
                return message;
            }
        }

        // Fallback: usar mensaje del enum
        return UtilsMessageFallback[code] || code;
    }

    /**
     * Obtiene el mensaje formateado con parametros
     * @param code Codigo de mensaje
     * @param params Parametros para formatear ({0}, {1}, etc.)
     * @param langId ID del idioma (opcional)
     * @returns Mensaje formateado
     */
    async getMessageWithParams(
        code: UtilsMessageCode,
        params: (string | number)[],
        langId?: LanguageId
    ): Promise<string> {
        const message = await this.getMessage(code, langId);
        return this.formatMessage(message, params);
    }

    /**
     * Obtiene el mensaje de forma sincrona (usa fallback)
     * Util para casos donde no se puede usar async/await
     * @param code Codigo de mensaje
     * @returns Mensaje fallback
     */
    getMessageSync(code: UtilsMessageCode): string {
        return UtilsMessageFallback[code] || code;
    }

    /**
     * Obtiene el mensaje formateado de forma sincrona
     * @param code Codigo de mensaje
     * @param params Parametros para formatear
     * @returns Mensaje formateado
     */
    getMessageSyncWithParams(code: UtilsMessageCode, params: (string | number)[]): string {
        const message = this.getMessageSync(code);
        return this.formatMessage(message, params);
    }

    /**
     * Crea una UtilsException con el codigo de mensaje
     * @param code Codigo de mensaje
     * @param additionalInfo Informacion adicional (opcional)
     * @returns UtilsException lista para lanzar
     */
    createException(code: UtilsMessageCode, additionalInfo?: string): UtilsException {
        const message = this.getMessageSync(code);
        this.logException(code, message, additionalInfo);
        return new UtilsException(code, message, additionalInfo);
    }

    /**
     * Crea una UtilsException con parametros formateados
     * @param code Codigo de mensaje
     * @param params Parametros para formatear
     * @param additionalInfo Informacion adicional (opcional)
     * @returns UtilsException lista para lanzar
     */
    createExceptionWithParams(
        code: UtilsMessageCode,
        params: (string | number)[],
        additionalInfo?: string
    ): UtilsException {
        const message = this.getMessageSyncWithParams(code, params);
        this.logException(code, message, additionalInfo);
        return new UtilsException(code, message, additionalInfo);
    }

    /**
     * Crea una UtilsException con causa original
     * @param code Codigo de mensaje
     * @param cause Error original
     * @param additionalInfo Informacion adicional (opcional)
     * @returns UtilsException lista para lanzar
     */
    createExceptionWithCause(
        code: UtilsMessageCode,
        cause: Error,
        additionalInfo?: string
    ): UtilsException {
        const message = this.getMessageSync(code);
        this.logException(code, message, additionalInfo, cause);
        return new UtilsException(code, message, additionalInfo, cause);
    }

    /**
     * Lanza una UtilsException con el codigo de mensaje
     * @param code Codigo de mensaje
     * @param additionalInfo Informacion adicional (opcional)
     * @throws UtilsException
     */
    throwException(code: UtilsMessageCode, additionalInfo?: string): never {
        throw this.createException(code, additionalInfo);
    }

    /**
     * Lanza una UtilsException con parametros formateados
     * @param code Codigo de mensaje
     * @param params Parametros para formatear
     * @param additionalInfo Informacion adicional (opcional)
     * @throws UtilsException
     */
    throwExceptionWithParams(
        code: UtilsMessageCode,
        params: (string | number)[],
        additionalInfo?: string
    ): never {
        throw this.createExceptionWithParams(code, params, additionalInfo);
    }

    /**
     * Formatea un mensaje reemplazando placeholders {0}, {1}, etc.
     * @param message Mensaje base
     * @param params Parametros para reemplazar
     * @returns Mensaje formateado
     */
    private formatMessage(message: string, params: (string | number)[]): string {
        if (!params || params.length === 0) {
            return message;
        }

        let result = message;
        params.forEach((param, index) => {
            const placeholder = `{${index}}`;
            result = result.replace(new RegExp(placeholder.replace(/[{}]/g, '\\$&'), 'g'), String(param));
        });

        return result;
    }

    /**
     * Registra el log de la excepcion segun el tipo de error
     * BUS### (negocio) -> console.warn
     * ERR### (tecnico) -> console.error
     */
    private logException(
        code: UtilsMessageCode,
        message: string,
        additionalInfo?: string,
        cause?: Error
    ): void {
        const isBusinessError = code.startsWith('BUS');
        const logMessage = additionalInfo
            ? `[${code}] ${message} - ${additionalInfo}`
            : `[${code}] ${message}`;

        if (isBusinessError) {
            console.warn(`[MessageCatalog] ${logMessage}`);
        } else if (cause) {
            console.error(`[MessageCatalog] ${logMessage}`, cause);
        } else {
            console.error(`[MessageCatalog] ${logMessage}`);
        }
    }

    /**
     * Indica si catalogos-api esta habilitado
     */
    isEnabled(): boolean {
        return catalogosConfig.enabled;
    }

    /**
     * Limpia el cache del cliente
     */
    clearCache(): void {
        catalogosClient.clearCache();
    }
}

// Singleton
export const messageCatalog = new MessageCatalogService();
export default messageCatalog;
